cumulusci/tasks/bulkdata/generate_and_load_data.py

```diff
@@ -0,0 +1,105 @@
+            "description": "Fully qualified class path of a task to generate the data. Use cumulusci.tasks.bulkdata.factory_generator if you would like to use a Factory Module.",
+            "required": False,
+        },
+        "data_generation_options": {
+            "description": "Options to pass to the data generator.",
+            "required": False,
+        },
+        "database_url": {
+            "description": "A URL to store the database (defaults to a transient SQLite file)",
+            "required": "",
+        },
+    }
+
+    def _run_task(self):
+        mapping_file = os.path.abspath(self.options["mapping"])
+        assert os.path.exists(mapping_file), f"{mapping_file} cannot be found."
+        database_url = self.options.get("database_url")
+        num_records = int(self.options["num_records"])
+        batch_size = int(self.options.get("batch_size", num_records))
+        if database_url and batch_size != num_records:
+            raise TaskOptionsError(
+                "You may not specify both `database_url` and `batch_size` options."
+            )
+
+        with temporary_dir() as tempdir:
+            for current_batch_size, index in self._batches(num_records, batch_size):
+                self._generate_batch(
+                    database_url, tempdir, mapping_file, current_batch_size, index
+                )
+
+    @staticmethod
+    def _batches(num_records, batch_size):
+        num_batches = (num_records // batch_size) + 1
+        for i in range(0, num_batches):
+            if i == num_batches - 1:  # last batch
+                batch_size = num_records - (batch_size * i)  # leftovers
+            if batch_size > 0:
+                yield batch_size, i
+
+    def _datagen(self, subtask_options):
+        class_path = self.options.get("data_generation_task", None)
+        task_class = import_global(class_path)
+        task_config = TaskConfig({"options": subtask_options})
+        data_gen_task = task_class(
+            self.project_config, task_config, org_config=self.org_config
+        )
+        data_gen_task()
+
+    def _dataload(self, subtask_options):
+        subtask_config = TaskConfig({"options": subtask_options})
+        subtask = LoadData(
+            project_config=self.project_config,
+            task_config=subtask_config,
+            org_config=self.org_config,
+            flow=self.flow,
+            name=self.name,
+            stepnum=self.stepnum,
+        )
+        subtask()
+
+    def _generate_batch(self, database_url, tempdir, mapping_file, batch_size, index):
+        if not database_url:
+            sqlite_path = os.path.join(tempdir, f"generated_data_{index}.db")
+            database_url = f"sqlite:///" + sqlite_path
+        subtask_options = {
+            **self.options,
+            "mapping": mapping_file,
+            "database_url": database_url,
+            "num_records": batch_size,
+        }
+        self._datagen(subtask_options)
```

Comment:
It seems like we should be passing the index into the data generator, in case it's important in order to generate a full sequence that spans multiple batches or something like that.

Reply:
https://github.com/SFDO-Tooling/CumulusCI/blob/f0eaab0c77ced04550e135113c2d06d4dbb50bcf/cumulusci/tasks/bulkdata/generate_and_load_data.py#L96

Reply:
Okay, the index is now passed around pervasively.